File: .NET/Microsoft.Recognizers.Text.DateTime/Utilities/MatchingUtil.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

using Microsoft.Recognizers.Text.Matcher;
using Microsoft.Recognizers.Text.Utilities;

namespace Microsoft.Recognizers.Text.DateTime
{
    public static class MatchingUtil
    {
        private const RegexOptions RegexFlags = RegexOptions.Singleline | RegexOptions.ExplicitCapture;
        private static readonly Regex InvalidDayNumberPrefix =
            new Regex(Definitions.BaseDateTime.InvalidDayNumberPrefix, RegexFlags);

        public static bool IsInvalidDayNumberPrefix(string prefix)
        {
            return InvalidDayNumberPrefix.IsMatch(prefix);
        }

        public static bool GetAgoLaterIndex(string text, Regex regex, out int index, bool inSuffix)
        {
            index = -1;
            var match = inSuffix ? regex.MatchBegin(text, trim: true) : regex.MatchEnd(text, trim: true);

            if (match.Success)
            {
                index = match.Index + (inSuffix ? match.Length : 0);
                return true;
            }

            return false;
        }

        public static bool GetTermIndex(string text, Regex regex, out int index)
        {
            index = -1;
            var match = regex.Match(text.Trim().Split(' ').Last());
            if (match.Success)
            {
                index = text.Length - text.LastIndexOf(match.Value, StringComparison.Ordinal);
                return true;
            }

            return false;
        }

        public static bool ContainsAgoLaterIndex(string text, Regex regex, bool inSuffix)
        {
            return GetAgoLaterIndex(text, regex, out var index, inSuffix);
        }

        public static bool ContainsTermIndex(string text, Regex regex)
        {
            return GetTermIndex(text, regex, out var index);
        }

        // Temporary solution for remove superfluous words only under the Preview mode
        public static string PreProcessTextRemoveSuperfluousWords(string text, StringMatcher matcher, out List<MatchResult<string>> superfluousWordMatches)
        {
            superfluousWordMatches = RemoveSubMatches(matcher.Find(text));

            var bias = 0;

            foreach (var match in superfluousWordMatches)
            {
                text = text.Remove(match.Start - bias, match.Length);
                bias += match.Length;
            }

            return text;
        }

        // Temporary solution for recover superfluous words only under the Preview mode
        public static List<ExtractResult> PostProcessRecoverSuperfluousWords(List<ExtractResult> extractResults, List<MatchResult<string>> superfluousWordMatches, string originText)
        {
            foreach (var match in superfluousWordMatches)
            {
                foreach (var extractResult in extractResults)
                {
                    var extractResultEnd = extractResult.Start + extractResult.Length;
                    if (match.Start > extractResult.Start && extractResultEnd >= match.Start)
                    {
                        extractResult.Length += match.Length;
                    }

                    if (match.Start <= extractResult.Start)
                    {
                        extractResult.Start += match.Length;
                    }
                }
            }

            foreach (var extractResult in extractResults)
            {
                extractResult.Text = originText.Substring((int)extractResult.Start, (int)extractResult.Length);
            }

            return extractResults;
        }

        public static List<MatchResult<string>> RemoveSubMatches(IEnumerable<MatchResult<string>> matchResults)
        {
            var matchList = matchResults.ToList();

            return matchList.Where(item =>
                !matchList.Any(
                    ritem => (ritem.Start < item.Start && ritem.End >= item.End) ||
                             (ritem.Start <= item.Start && ritem.End > item.End))).ToList();
        }
    }
}

File: Python/libraries/recognizers-date-time/recognizers_date_time/date_time/utilities/matching_util.py
from typing import List, Pattern
import regex

from recognizers_text.extractor import ExtractResult
from recognizers_text.utilities import RegExpUtility
from recognizers_text.matcher.string_matcher import MatchResult
from recognizers_date_time.resources.base_date_time import BaseDateTime


class MatchedIndex:
    def __init__(self, matched: bool, index: int):
        self.matched = matched
        self.index = index


class MatchingUtil:

    invalid_day_number_prefix = RegExpUtility.get_safe_reg_exp(
        BaseDateTime.InvalidDayNumberPrefix)

    @staticmethod
    def is_invalid_day_number_prefix(prefix: str) -> bool:
        return MatchingUtil.invalid_day_number_prefix.search(prefix)

    @staticmethod
    def pre_process_text_remove_superfluous_words(text: str, matcher: Pattern):
        superfluous_word_matches = MatchingUtil.remove_sub_matches(matcher.find(text))

        bias = 0

        for match in superfluous_word_matches:
            text = text[match.start - bias: match.length]

            bias += match.length

        return text, superfluous_word_matches

    @staticmethod
    def post_process_recover_superfluous_words(extract_results: List[ExtractResult], superfluous_word_matches,
                                               origin_text: str):
        for match in superfluous_word_matches:
            for extract_result in extract_results:

                extract_result_end = extract_result.start + extract_result.length
                if extract_result.start < match.start <= extract_result_end:
                    extract_result.length += len(match)

                if match.start <= extract_result.start:
                    extract_result.start += len(match)

        for extract_result in extract_results:
            extract_result.text = origin_text[extract_result.start: extract_result.start + extract_result.length]

        return extract_results

    @staticmethod
    def remove_sub_matches(match_results: List[MatchResult]):
        match_list = list(match_results)

        match_list = (list(filter(lambda item: not any(list(filter(
            lambda ritem: (ritem.start < item.start and ritem.end >= item.end) or (
                ritem.start <= item.start and ritem.end > item.end), match_list))), match_list)))

        return match_list

    @staticmethod
    def get_ago_later_index(source: str, regexp: Pattern, in_suffix) -> MatchedIndex:
        result = MatchedIndex(matched=False, index=-1)
        trimmed_source = source.strip().lower()
        match = RegExpUtility.match_begin(regexp, trimmed_source, True) if in_suffix else\
            RegExpUtility.match_end(regexp, trimmed_source, True)

        if match and match.success:
            result.index = source.lower().find(match.group()) + (match.length if in_suffix else 0)
            result.matched = True

        return result

    @staticmethod
    def contains_ago_later_index(source: str, regexp: Pattern, in_suffix) -> bool:
        return MatchingUtil.get_ago_later_index(source, regexp, in_suffix).matched

    @staticmethod
    def get_term_index(source: str, regexp: Pattern) -> MatchedIndex:
        result = MatchedIndex(matched=False, index=-1)
        referenced_match = regex.search(
            regexp, source.strip().lower().split(' ').pop())

        if referenced_match:
            result = MatchedIndex(matched=True, index=len(
                source) - source.lower().rfind(referenced_match.group()))

        return result

    @staticmethod
    def contains_term_index(source: str, regexp: Pattern) -> bool:
        return MatchingUtil.get_term_index(source, regexp).matched